bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_open(std::move(end));
+  return tmp;
+}
+
+/**
+ * open interval: (begin, end), both endpoints should be set.
+ */
+RowRange RowRange::open_interval(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_open(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_open(std::move(end));
+  return tmp;
+}
+
+// closed interval: [begin, end], both endpoints should be set.
+RowRange RowRange::closed_interval(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_closed(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_closed(std::move(end));
+  return tmp;
+}
+
+// left-open: (begin, end], both endpoints should be set.
+RowRange RowRange::left_open(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_open(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_closed(std::move(end));
+  return tmp;
+}
+
+/**
+ * Given any key 'prefix', the prefix range is given as the right-open interval:
+ *   ['prefix', successor('prefix')).
+ *
+ * When prefix is of the form "\xFF\xFF\xFF...\xFF", the successor is
+ * "" (empty string). In such a case, we set end key to 'prefix';
+ *
+ */
+RowRange RowRange::prefix_range(std::string prefix) {
+  RowRange tmp;
+  std::string next_prefix = Successor(prefix);
+
+  if (next_prefix.empty()) next_prefix = prefix;
+  tmp.range_.set_start_key_closed(std::move(prefix));
+  if (!next_prefix.empty())  // Infinite range for prefix="".
+    tmp.range_.set_end_key_open(std::move(next_prefix));
+  return tmp;
+}
+
+// An alias for right_open().
+RowRange RowRange::range(std::string begin, std::string end) {
+  return right_open(begin, end);
+}
+
+/*
+ * Increment the given key: gives the permutation string (in the key-space)
+ * that immediately follows the 'key'.
+ * Please note that, for a key of the form "\xFF\xFF\xFF...\xFF", the successor
+ * is "" (empty string).
+ */
+std::string RowRange::Successor(std::string const& key) {
+  if (key.empty()) {
+    return "";
+  }
+  std::size_t position = key.find_last_not_of('\xFF');
+
+  if (position == std::string::npos) {
+    return "";
+  }
+  const char last_char = key[position] + 1;
+  return (key.substr(0, position) + last_char);
+}
+
+/**
+ * A range  can be of folling types:
+ *  - start is NOT_SET or set to "",  and end is NOT_SET: infinite range
+ *    spanning the complete domain of the row key.
+ *  - start key set to a non-empty 'x', and end is NOT_SET: one-sided infinite
+ *    range starting from x.
+ */
+bool RowRange::contains(std::string_view row) const {
+  KeyType start_type{GetStartKeyType()}, end_type{GetEndKeyType()};
+  std::string start_key{GetStartKey()}, end_key{GetEndKey()};
+  bool result = true;
+
+  if (row.empty()) return true;
```

Comment:
Hmmm, I would be tempted to return false...

The algorithm to decide if something is inside a range should be simpler than this.  Basically, a thing is inside a range if it is not outside the range.  So:

```C++
bool RowRange::contains(absl::string_view key) const {
  if (before_begin(key)) {
    return false;
  }
  if (after_end(key)) {
    return false;
  }
  return true;
}

// This should be private:
bool RowRange::before_begin(absl::string_view key) const {
  switch (range_.start_key_case()) {
    case btproto::RowRange::kStartKeyOpen:
      return key <= range_.start_key_open();
    case btproto::RowRange::kStartKeyClosed:
      return key < range_.start_key_closed();
    case btproto::RowRange::START_KEY_NOT_SET:
      // FALLTHROUGH
  }
  return false;
}

// This should be private:
bool RowRange::after_end(absl::string_view key) const {
// mutatis mutandi from before_begin
}
```

Reply:
How about this? Using methods that are more general and work on both points, and these two methods can  support other two range inequalities  (>, >=), in case we need in future? 
```
bool RowRange::contains(absl::string_view key) const {
  return  (!less_than(key, PositionType::START) && 
                less_than_equal(key, PositionType::END));
  }   
```